File: Pythoncode/ARIMA4.py
# 該当ＵＲＬを変数に格納
repo = "https://raw.githubusercontent.com/u-10bei/Population_Projections/"
popURL = repo + "main/data/population_jp_year.csv"

# ライブラリの読み込み
import pandas as pd
from dateutil.relativedelta import relativedelta

# ネット上のファイル読み込み
pop_csv = pd.read_csv( popURL,
                      parse_dates = [ 'Year' ])
pop_csv.Year = pop_csv.Year.apply( lambda x: x + relativedelta( months = 9 ))
pop_df = pop_csv[[ 'Year', 'Total' ]]\
         .set_index( 'Year' )
pop_dfB = pop_csv[[ 'Year', 'Birth' ]]\
          .astype({ 'Birth': float })\
          .set_index( 'Year' )
pop_dfD = pop_csv[[ 'Year', 'Tu14', 'Du14', 'Tm', 'Dm', 'To65', 'Do65' ]]\
          .set_index( 'Year' )
pop_dfDru = pop_dfD[[ 'Du14' ]]\
          .astype({ 'Du14': float })\
          .rename(columns={ 'Du14': 'Dru' })
pop_dfDrm = pop_dfD[[ 'Dm' ]]\
          .astype({ 'Dm': float })\
          .rename(columns={ 'Dm': 'Drm' })
pop_dfDro = pop_dfD[[ 'Do65' ]]\
          .astype({ 'Do65': float })\
          .rename(columns={ 'Do65': 'Dro' })
pop_dfDru.Dru = pop_dfD.Du14 / pop_dfD.Tu14
pop_dfDrm.Drm = pop_dfD.Dm / pop_dfD.Tm
pop_dfDro.Dro = pop_dfD.Do65 / pop_dfD.To65

# ライブラリの読み込み
import matplotlib.pyplot as plt
plt.style.use( 'ggplot' )
plt.plot( pop_dfB )

fig = plt.figure( figsize = [ 16, 5 ])

ax1 = fig.add_subplot( 1, 3, 1 )   #１行３列の１番目
ax1.plot( pop_dfDru )

ax2 = fig.add_subplot( 1, 3, 2 )   #１行３列の２番目
ax2.plot( pop_dfDrm )

ax3 = fig.add_subplot( 1, 3, 3 )   #１行３列の３番目
ax3.plot( pop_dfDro )

fig.tight_layout()
plt.show()

# ライブラリの読み込み
import statsmodels.api as sm
sm.graphics.tsa.plot_acf( pop_dfB,
                          lags = 20,
                          alpha = .05,
                          title = "Birth" )
plt.show()

fig = plt.figure( figsize = [ 16, 5 ])

ax1 = fig.add_subplot( 1, 3, 1 )   #１行３列の１番目
# 自己相関のグラフ
sm.graphics.tsa.plot_acf( pop_dfDru,
                          lags = 20,
                          alpha = .05,
                          ax = ax1,
                          title = "Death under14" )

ax2 = fig.add_subplot( 1, 3, 2 )   #１行３列の２番目
sm.graphics.tsa.plot_acf( pop_dfDrm,
                          lags = 20,
                          alpha = .05,
                          ax = ax2,
                          title = "Death 15 to 64" )

ax3 = fig.add_subplot( 1, 3, 3 )   #１行３列の３番目
sm.graphics.tsa.plot_acf( pop_dfDro,
                          lags = 20,
                          alpha = .05,
                          ax = ax3,
                          title = "Death over65" )

fig.tight_layout()
plt.show()

sm.graphics.tsa.plot_pacf( pop_dfB,
                          lags = 20,
                          alpha = .05,
                          title = "Birth" )
plt.show()

fig = plt.figure( figsize = [ 16, 5 ])

ax1 = fig.add_subplot( 1, 3, 1 )   #１行３列の１番目
# 偏自己相関のグラフ
sm.graphics.tsa.plot_pacf( pop_dfDru,
                          lags = 20,
                          alpha = .05,
                          ax = ax1,
                          title = "Death under14" )

ax2 = fig.add_subplot( 1, 3, 2 )   #１行３列の２番目
sm.graphics.tsa.plot_pacf( pop_dfDrm,
                          lags = 20,
                          alpha = .05,
                          ax = ax2,
                          title = "Death 15 to 64" )

ax3 = fig.add_subplot( 1, 3, 3 )   #１行３列の３番目
sm.graphics.tsa.plot_pacf( pop_dfDro,
                          lags = 20,
                          alpha = .05,
                          ax = ax3,
                          title = "Death over65" )

fig.tight_layout()
plt.show()

# 学習データと予測データ
prow_test = 6
pop_test2 = pop_df.tail( prow_test )
pop_testD = pop_dfD.tail( prow_test )
pop_testB = pop_dfB.tail( prow_test )
pop_trainB = pop_dfB.head( prow_test * -1 )
pop_testDru = pop_dfDru.tail( prow_test )
pop_trainDru = pop_dfDru.head( prow_test * -1 )
pop_testDrm = pop_dfDrm.tail( prow_test )
pop_trainDrm = pop_dfDrm.head( prow_test * -1 )
pop_testDro = pop_dfDro.tail( prow_test )
pop_trainDro = pop_dfDro.head( prow_test * -1 )

# ライブラリの読み込み
import pmdarima as pm

# ＡＲＩＭＡモデルの推定
pop_arimaB = pm.auto_arima( pop_trainB,
                           information_criterion = 'aic' )
pop_arimaDru = pm.auto_arima( pop_trainDru,
                           information_criterion = 'aic' )
pop_arimaDrm = pm.auto_arima( pop_trainDrm,
                           information_criterion = 'aic' )
pop_arimaDro = pm.auto_arima( pop_trainDro,
                           information_criterion = 'aic' )
pop_arimaB,pop_arimaDru,pop_arimaDrm,pop_arimaDro

# モデリング
from statsmodels.tsa.arima.model import ARIMA
pop_arimaB = ARIMA( pop_trainB, order = ( 1, 1, 1 )).fit()
pop_arimaDru = ARIMA( pop_trainDru, order = ( 4, 2, 0 )).fit()
pop_arimaDrm = ARIMA( pop_trainDrm, order = ( 0, 2, 1 )).fit()
pop_arimaDro = ARIMA( pop_trainDro, order = ( 0, 1, 1 )).fit()

# ＡＲＩＭＡによる予測
pop_arima_fB = pop_arimaB.predict( '2015', '2020' )
pop_arima_fDru = pop_arimaDru.predict( '2015', '2020' )
pop_arima_fDrm = pop_arimaDrm.predict( '2015', '2020' )
pop_arima_fDro = pop_arimaDro.predict( '2015', '2020' )
pop_arima_fB,pop_arima_fDru,pop_arima_fDrm,pop_arima_fDro

pop_Total = pop_test2.Total +\
            pop_arima_fB -\
            ( pop_testD.Tu14 *\
             pop_arima_fDru ) -\
            ( pop_testD.Tm *\
             pop_arima_fDrm ) -\
            ( pop_testD.To65 *\
             pop_arima_fDro )
pop_arima_f4 = pop_Total.shift( 1 ).tail()

# 社人研予測との比較
# 該当ＵＲＬを変数に格納
ipssURL = repo + "main/data/forecast_ipss.csv"

# ネット上のファイル読み込み
ipss_csv = pd.read_csv( ipssURL,
                        parse_dates = [ 'Year' ])
ipss_csv.Year = ipss_csv.Year.apply( lambda x: x + relativedelta( months = 9 ))
ipss_test = ipss_csv.set_index( 'Year' )

# ライブラリの読み込み
import numpy as np

plot_test2 = pop_test2.tail()
# 予測と実測の比較（グラフ）
x_axis = np.arange(pop_arima_f4.shape[0])
plt.figure( figsize = [ 16, 5 ])
plt.plot( x_axis,
          plot_test2,
          label = "Total",
          color = 'black' )
plt.plot( x_axis,
          pop_arima_f4,
          label="forecast",
          color='blue')
plt.plot( x_axis,
          ipss_test.DMBM,
          label="DMBM",
          color='green')
plt.plot( x_axis,
          ipss_test.DMBH,
          label="DMBH",
          color='green')
plt.plot( x_axis,
          ipss_test.DLBH,
          label="DLBH",
          color='green')
plt.plot( x_axis,
          ipss_test.DLBM,
          label="DLBM",
          color='red')
plt.legend()
plt.show()

File: Rcode/ARIMA4.R
#ライブラリインストール(必要に応じて)
#install.packages( "tidyverse" )
#install.packages( "fable" )
#install.packages( "feasts" )
#install.packages( "urca" )
#install.packages( "reshape2" )

# 該当リポジトリを変数に格納
c( "https://raw.githubusercontent.com/u-10bei/Population_Projections/" ) -> repo
# 該当ＵＲＬを変数に格納
repo |> paste0( c( "main/data/population_jp_year.csv" )) -> popURL

# ライブラリの読み込み
library( readr )
library( fable )
library( dplyr )

# ネット上のファイル読み込み
popURL |>
  read_csv( show_col_types = FALSE ) |>
  # ＴＳＩＢＢＬＥライブラリに変換
  as_tsibble( index = Year ) |>
  mutate( Dr = Death / Total,
          Dru14 = Du14 / Tu14,
          Drm = Dm / Tm,
          Dro65 = Do65 / To65 ) -> pop_tsibble

# ライブラリの読み込み
library( ggplot2 )

# 出生数、死亡率のグラフ
pop_tsibble |>
  autoplot( Birth )
pop_tsibble |>
  autoplot( Dr )
pop_tsibble |>
  autoplot( Dru14 )
pop_tsibble |>
  autoplot( Drm )
pop_tsibble |>
  autoplot( Dro65 )

# ライブラリの読み込み
library( feasts )

# 自己相関のグラフ
pop_tsibble |>
  ACF( Birth ) |>
  autoplot()
pop_tsibble |>
  ACF( Dr ) |>
  autoplot()
pop_tsibble |>
  ACF( Dru14 ) |>
  autoplot()
pop_tsibble |>
  ACF( Drm ) |>
  autoplot()
pop_tsibble |>
  ACF( Dro65 ) |>
  autoplot()

# 偏自己相関のグラフ
pop_tsibble |>
  PACF( Birth ) |>
  autoplot()
pop_tsibble |>
  PACF( Dr ) |>
  autoplot()
pop_tsibble |>
  PACF( Dru14 ) |>
  autoplot()
pop_tsibble |>
  PACF( Drm ) |>
  autoplot()
pop_tsibble |>
  PACF( Dro65 ) |>
  autoplot()

pop_tsibble |>
  model(STL( Birth ~ season( window = Inf ))) |>
  components() |>
  autoplot()
pop_tsibble |>
  model(STL( Dr ~ season( window = Inf ))) |>
  components() |>
  autoplot()
pop_tsibble |>
  model(STL( Dru14 ~ season( window = Inf ))) |>
  components() |>
  autoplot()
pop_tsibble |>
  model(STL( Drm ~ season( window = Inf ))) |>
  components() |>
  autoplot()
pop_tsibble |>
  model(STL( Dro65 ~ season( window = Inf ))) |>
  components() |>
  autoplot()

# 学習データと予測データ
6 -> prow_test2
pop_tsibble |> nrow() - prow_test2 -> prow_train2
pop_tsibble |> tail( n = prow_test2 ) -> pop_test2
pop_tsibble |> head( n = prow_train2 ) -> pop_train2

# ＡＲＩＭＡモデルの推定
pop_train2 |>
  model( arima = ARIMA( Birth,
                        ic = "aic",
                        stepwise = FALSE )) -> pop_arimaB
pop_train2 |>
  model( arima = ARIMA( Dru14,
                        ic = "aic",
                        stepwise = FALSE )) -> pop_arimaDru
pop_train2 |>
  model( arima = ARIMA( Drm,
                        ic = "aic",
                        stepwise = FALSE )) -> pop_arimaDrm
pop_train2 |>
  model( arima = ARIMA( Dro65,
                        ic = "aic",
                        stepwise = FALSE )) -> pop_arimaDro

# ＡＲＩＭＡによる予測
pop_arimaB |>
forecast( xreg = pop_test2$Birth,
          h = "6 years") -> pop_arimaB_f
pop_arimaDru |>
  forecast( xreg = pop_test2$Dru14,
            h = "6 years") -> pop_arimaDru_f
pop_arimaDrm |>
  forecast( xreg = pop_test2$Drm,
            h = "6 years") -> pop_arimaDrm_f
pop_arimaDro |>
  forecast( xreg = pop_test2$Dro65,
            h = "6 years") -> pop_arimaDro_f

# 社人研予測との比較
# 該当ＵＲＬを変数に格納
repo |> paste0( c( "main/data/forecast_ipss.csv" )) -> ipssURL

# ネット上のファイル読み込み
ipssURL |>
  read_csv( show_col_types = FALSE ) |>
  # ＴＳＩＢＢＬＥライブラリに変換
  as_tsibble( index = Year ) -> ipss_test

# 出生数、死亡数の合算
pop_test2 |> rename( "forecast_BD" = Total ) -> pop_arima_f4

pop_arimaB_f |>
  as.data.frame() |>
  select( .mean ) -> pop_arima_f4[ ,3 ]
pop_arimaDru_f |>
  as.data.frame() |>
  select( .mean ) -> pop_arima_f4[ ,12 ]
pop_arimaDrm_f |>
  as.data.frame() |>
  select( .mean ) -> pop_arima_f4[ ,13 ]
pop_arimaDro_f |>
  as.data.frame() |>
  select( .mean ) -> pop_arima_f4[ ,14 ]
pop_arima_f4 |>
  mutate( Du14 = Tu14  * Dru14,
          Dm = Tm * Drm,
          Do65 = To65 * Dro65,
          forecast_BD = lag( forecast_BD +
                               Birth -
                               Du14 -
                               Dm - 
                               Do65 )) -> pop_arima_f4

pop_arima_f4[ 2:6, 1:2 ] |>
  inner_join( pop_test, by = "Year") |>
  inner_join( ipss_test, by = "Year") |>
  select( Year,
          Total,
          forecast_BD,
          DMBM,
          DMBH,
          DLBM,
          DLBH ) -> join_test4

# ライブラリの読み込み
library( reshape2 )

join_test4 |> 
  melt(id="Year",measure=c( "Total",
                            "forecast_BD",
                            "DMBM",
                            "DMBH",
                            "DLBM",
                            "DLBH")) -> join_plot4

#描画
ggplot( join_plot4,
        aes(x = Year,
            y = value,
            shape = variable,
            colour = variable,
            group = variable )) +
  geom_line() +
  geom_point()

pop_test2 |>
  select( Year, Birth ) -> pop_testB
pop_arimaB_f |>
  autoplot() +
  autolayer( pop_testB )
pop_test2 |>
  select( Year, Dru14 ) -> pop_testDru
pop_arimaDru_f |>
  autoplot() +
  autolayer( pop_testDru )
pop_test2 |>
  select( Year, Drm ) -> pop_testDrm
pop_arimaDrm_f |>
  autoplot() +
  autolayer( pop_testDrm )
pop_test2 |>
  select( Year, Dro65 ) -> pop_testDro
pop_arimaDro_f |>
  autoplot() +
  autolayer( pop_testDro )
